blog: Announcing Flux 2.7 GA (#2365)
fluxcd/website: content/en/blog/2025-09-29-announcing-flux-v2.7.0/index.md

- `Provider.spec.serviceAccountName` for authenticating with Azure DevOps, Azure Event Hub and Google Pub/Sub.

For more details on how to configure object-level workload identity for Flux, see the following docs:

- [AWS workload identity](/flux/integrations/aws/)
- [Azure workload identity](/flux/integrations/azure/)
- [GCP workload identity](/flux/integrations/gcp/)

## OpenTelemetry Tracing

Starting with Flux v2.7, users can enable OpenTelemetry tracing for Flux reconciliations
by configuring a Provider of type `otel`:

```yaml
apiVersion: notification.toolkit.fluxcd.io/v1beta3
kind: Provider
metadata:
  name: jaeger
  namespace: flux-system
spec:
  type: otel
  address: http://jaeger-collector.jaeger:4318/v1/traces
```

The notification-controller converts Flux events into OTEL spans with proper trace relationships
based on the Flux object hierarchy. Source objects (`GitRepository`, `HelmChart`, `OCIRepository`, `Bucket`)
create root spans, while `Kustomization`and `HelmRelease` objects create child spans within the same trace.
Each span includes event metadata as attributes and uses the alert name and namespace as the service identifier.

> ![fluxcd-otel](flux-otel-trace.png)

Reviewer: Everything is really tiny in this screenshot. It's nearly impossible to recognize anything.

Author: Right, you need to zoom in on your browser to see what's going on. There must be a Hugo plugin that shown images full screen...

Reviewer: There is: https://github.com/fluxcd/website/blob/18fccdbcdd61bc19885f3717f7a08da1909b3675/content/en/blog/2023-08-28-announcing-flux-v2.1.0/index.md#flux-grafana-dashboards

This might help seeing better what's going on.

leaving other services untouched.

For more details on how to use the `ArtifactGenerator` API, please refer to the
[source-watcher documentation](/flux/components/source/artifactgenerators/).

## Supported Versions

Flux v2.4 has reached end-of-life and is no longer supported.

Flux v2.7 supports the following Kubernetes versions:

| Distribution | Versions         |
|:-------------|:-----------------|
| Kubernetes   | 1.32, 1.33, 1.34 |
| OpenShift    | 4.19             |

> **Enterprise support** Note that the CNCF Flux project offers support only for the latest three minor versions of Kubernetes.
> Backwards compatibility with older versions of Kubernetes and OpenShift is offered by vendors such as [ControlPlane](https://control-plane.io/enterprise-for-flux-cd/) that provide enterprise support for Flux.

## Upgrade Procedure

Note that in Flux v2.7, the following APIs have reached end-of-life and have being removed from the CRDs:

- `source.toolkit.fluxcd.io/v1beta1`
- `kustomize.toolkit.fluxcd.io/v1beta1`
- `helm.toolkit.fluxcd.io/v2beta1`
- `image.toolkit.fluxcd.io/v1beta1`
- `notification.toolkit.fluxcd.io/v1beta1`

> ### Step1: Migrate to stable APIs in Git

Reviewer: ```suggestion
### Step 1: Migrate to stable APIs in Git
```

Migrated all your resources to the Flux v2.6 stable APIs in your Git repositories by replacing
the API version in the manifests:

- `Kustomization` to `kustomize.toolkit.fluxcd.io/v1`
- `HelmRelease` to `helm.toolkit.fluxcd.io/v2`
- `Bucket` to `source.toolkit.fluxcd.io/v1`
- `GitRepository` to `source.toolkit.fluxcd.io/v1`
- `HelmChart` to `source.toolkit.fluxcd.io/v1`
- `HelmRepository` to `source.toolkit.fluxcd.io/v1`
- `OCIRepository` to `source.toolkit.fluxcd.io/v1`
- `Receiver` to `notification.toolkit.fluxcd.io/v1`
- `Alert` to `notification.toolkit.fluxcd.io/v1beta3`
- `Provider` to `notification.toolkit.fluxcd.io/v1beta3`
- `ImageRepository` to `image.toolkit.fluxcd.io/v1beta2`
- `ImagePolicy` to `image.toolkit.fluxcd.io/v1beta2`
- `ImageUpdateAutomation` to `image.toolkit.fluxcd.io/v1beta2`

Note that the `ImageUpdateAutomation` commit template should use the fields
`.Changed.FileChanges`, `.Changed.Objects` and `.Changed.Changes` instead of the deprecated
`.Updated` and `.Changed.ImageResult` fields.

Commit and push the changes to your Git repositories, then wait for Flux v2.6 to reconcile the new API versions.

> ### Step2: Migrate to stable APIs in Kubernetes etcd

Reviewer: ```suggestion
### Step 2: Migrate to stable APIs in Kubernetes etcd
```

Download and install the latest Flux CLI v2.7 using the instructions
from [installation guide](/flux/installation/#install-the-flux-cli).

Then run the following command on all your clusters where Flux is installed:

```sh
flux migrate
```

The command migrates the Flux custom resources stored in Kubernetes etcd to their latest API version,
ensuring the Flux components can continue to function correctly after the 2.7 upgrade.

> Running the migrate command is required unless you are using
> the [Flux Operator](https://github.com/controlplaneio-fluxcd/flux-operator) which
> automatically handles the migration of Flux resources in etcd.

> ### Step3: Upgrade Flux components

Reviewer: ```suggestion
### Step 3: Upgrade Flux components
```

Upgrade the Flux components to v2.7 by running `flux bootstrap` on your clusters
with the same parameters you used during the initial installation. 

At this step, you can also enable the new `source-watcher` component
by adding the `--components-extra=source-watcher` flag.

> If you are using the [Flux Operator](https://github.com/controlplaneio-fluxcd/flux-operator), make sure you've
> updated the operator to v0.30 or later, set `FluxInstance.spec.distribution.version` to `2.7.x`
> and `FluxInstance.spec.components` to include `source-watcher`.

> ### Step4: Migrate the Image APIs in Git

Reviewer: ```suggestion
### Step 4: Migrate the Image APIs in Git
```

including watching for changes in ConfigMaps and Secrets references,
extended readiness evaluation of dependencies with CEL expressions,
and support for OpenTelemetry tracing for Flux `Kustomization` and `HelmRelease` reconciliation.

In ecosystem news, there is a new release of [Flux Operator](https://github.com/controlplaneio-fluxcd/flux-operator)
that comes with [in-cluster image update automation](https://fluxcd.control-plane.io/operator/resourcesets/image-automation/)
features, that can be used for GitLess GitOps workflows.

## General availability of Image Update Automation

This release marks the General Availability (GA) of Flux [Image Automation](/flux/components/image/)
APIs and controllers. The image-reflector-controller and image-automation-controller work together to
update Kubernetes manifests in Git repositories when new container images are available in
container registries.

The following APIs have been promoted to stable v1:

- [ImageRepository](/flux/components/image/imagerepositories/)
- [ImagePolicy](/flux/components/image/imagepolicies/)
- [ImageUpdateAutomation](/flux/components/image/imageupdateautomations/)

> The `ImagePolicy` API  now supports the `.spec.suspend` field to pause and resume the policy evaluation.

Reviewer 2: ```suggestion
The `ImagePolicy` API now supports the `.spec.suspend` field to pause and resume the policy evaluation.
```

Starting with Flux v2.7, the `kustomize-controller`, `helm-controller` and `notification-controller`
gain support for reacting to changes in ConfigMaps and Secrets references.

The following references are now watched for changes:

- `Kustomization.spec.postBuild.substituteFrom`
- `Kustomization.spec.decryption.secretRef`
- `Kustomization.spec.kubeConfig.secretRef`
- `Kustomization.spec.kubeConfig.configMapRef`
- `HelmRelease.spec.valuesFrom`
- `HelmRelease.spec.kubeConfig.secretRef`
- `HelmRelease.spec.kubeConfig.configMapRef`
- `Provider.spec.secretRef`

When a referenced ConfigMap or Secret changes, the controller will immediately trigger a reconciliation
if the referenced object is labelled with `reconcile.fluxcd.io/watch: Enabled`.

To enable the watching of all referenced objects without the need to label them,
the controllers can be configured with the `--watch-configs-label-selector=owner!=helm` flag.

## Workload Identity Authentication for Remote Clusters

Starting with Flux v2.7, you can configure workload identity at the object level
in the `Kustomization` and `HelmRelease` resources to authenticate with cloud providers

> when running Flux in the hub-and-spoke mode.

Reviewer 2: ```suggestion
when running Flux in the hub-and-spoke model.
```

- `Kustomization.spec.kubeConfig.configMapRef.name` for authenticating with remote clusters on AWS EKS, Azure AKS and Google GKE.
- `HelmRelease.spec.kubeConfig.configMapRef.name` for authenticating with remote clusters on AWS EKS, Azure AKS and Google GKE.
- `Provider.spec.serviceAccountName` for authenticating with Azure DevOps, Azure Event Hub and Google Pub/Sub.

For more details on how to configure object-level workload identity for Flux, see the following docs:

- [AWS workload identity](/flux/integrations/aws/)
- [Azure workload identity](/flux/integrations/azure/)
- [GCP workload identity](/flux/integrations/gcp/)

## OpenTelemetry Tracing

Starting with Flux v2.7, users can enable OpenTelemetry tracing for Flux reconciliations
by configuring a Provider of type `otel`:

```yaml
apiVersion: notification.toolkit.fluxcd.io/v1beta3
kind: Provider
metadata:
  name: jaeger
  namespace: flux-system
spec:
  type: otel
  address: http://jaeger-collector.jaeger:4318/v1/traces
```

The notification-controller converts Flux events into OTEL spans with proper trace relationships
based on the Flux object hierarchy. Source objects (`GitRepository`, `HelmChart`, `OCIRepository`, `Bucket`)

> create root spans, while `Kustomization`and `HelmRelease` objects create child spans within the same trace.

Reviewer 2: ```suggestion
create root spans, while `Kustomization` and `HelmRelease` objects create child spans within the same trace.
```

> Each span includes event metadata as attributes and uses the alert name and namespace as the service identifier.

Reviewer 2: @adri1197 Can you please review the wording here?

{{< gallery match="images/*" sortOrder="desc" rowHeight="150" margins="5"
thumbnailResizeOptions="900x900 q90 Lanczos"
previewType="blur" embedPreview=true lastRow="nojustify" >}}

For more details on how to configure OpenTelemetry tracing for Flux, please refer to the
[notification-controller documentation](/flux/components/notification/providers.md#otel).

## Controller Improvements

> - The `GitRepository` API gains supports mTLS for GitHub App authentication.

Reviewer 2: ```suggestion
- The `GitRepository` API gains support for mTLS in GitHub App authentication.
```

- The `Kustomization` API now supports [CEL expressions](/flux/components/kustomize/kustomizations.md#dependency-ready-expression) for extended readiness evaluation of dependencies.

> - The `Kustomization` API gains a new field `.spec.ignoreMissingComponents` for ignoring missing Kustomize components in source.

Reviewer 2: ```suggestion
- The `Kustomization` API gains a new field `.spec.ignoreMissingComponents` for ignoring missing Kustomize components in the source.
```

> - The kustomize-controller now supports global SOPS Age key decryption, allowing centralized management of decryption keys.

Reviewer 2: ```suggestion
- The kustomize-controller now supports global SOPS decryption for Age keys, allowing centralized management of decryption keys.
```

> - The kustomize-controller can be configured to cancel ongoing health checks when a new source revision is detected with the `--feature-gates=CancelHealthChecksOnNewRevision=true` flag.

Reviewer 2: ```suggestion
- The kustomize-controller can be configured to cancel ongoing health checks when a new source revision is detected with the `--feature-gates=CancelHealthCheckOnNewRevision=true` flag.
```

Reviewer 2: (The gate is `CancelHealthCheckOnNewRevision`, not `CancelHealthChecksOnNewRevision`)

- The `HelmRelease` API now supports [CEL expressions](/flux/components/helm/helmreleases.md#dependency-ready-expression) for extended readiness evaluation of dependencies.

> - The `HelmRelease` API gains a new strategy called `RetryOnFailure` for better handling of release upgrades failures.

Reviewer 2: ```suggestion
- The `HelmRelease` API gains a new strategy called `RetryOnFailure` for better handling of release failures.
```

- The `Provider` API now supports setting proxy via `spec.proxySecretRef` and mTLS via `spec.certSecretRef`.
- The `Provider` API has been extended with support for Zulip and OpenTelemetry tracing.

## Artifact Generators

> Flux v2.7 comes with a new component that be enabled at bootstrap time

Reviewer 2: ```suggestion
Flux v2.7 comes with a new component that can be enabled at bootstrap time
```

kind: Kustomization
metadata:
  name: backend-service
  namespace: apps
spec:
  interval: 15m
  prune: true
  sourceRef:
    kind: ExternalArtifact
    name: backend
  path: ./
```

Each service gets its own `ExternalArtifact` with an independent revision.
Changes to `deploy/backend/` only trigger the reconciliation of the backend-service `Kustomization`,
leaving other services untouched.

For more details on how to use the `ArtifactGenerator` API, please refer to the
[source-watcher documentation](/flux/components/source/artifactgenerators/).

## Supported Versions

Flux v2.4 has reached end-of-life and is no longer supported.

Flux v2.7 supports the following Kubernetes versions:

| Distribution | Versions         |
|:-------------|:-----------------|
| Kubernetes   | 1.32, 1.33, 1.34 |
| OpenShift    | 4.19             |

> **Enterprise support** Note that the CNCF Flux project offers support only for the latest three minor versions of Kubernetes.
> Backwards compatibility with older versions of Kubernetes and OpenShift is offered by vendors such as [ControlPlane](https://control-plane.io/enterprise-for-flux-cd/) that provide enterprise support for Flux.

## Upgrade Procedure

> Note that in Flux v2.7, the following APIs have reached end-of-life and have being removed from the CRDs:

Reviewer 2: ```suggestion
Note that in Flux v2.7, the following APIs have reached end-of-life and have been removed from the CRDs:
```

- `source.toolkit.fluxcd.io/v1beta1`
- `kustomize.toolkit.fluxcd.io/v1beta1`
- `helm.toolkit.fluxcd.io/v2beta1`
- `image.toolkit.fluxcd.io/v1beta1`
- `notification.toolkit.fluxcd.io/v1beta1`

### Step 1: Migrate to stable APIs in Git

> Migrated all your resources to the Flux v2.6 stable APIs in your Git repositories by replacing

Reviewer 2: ```suggestion
Migrate all your resources to the Flux v2.6 stable APIs in your Git repositories by replacing
```
